pkg/flavor/kubernetes/multiwriter.go

```diff
@@ -0,0 +1,673 @@
+	mo := req.GetVolumeCapability().GetMount().GetMountFlags()
+	if req.GetReadonly() {
+		mo = append(mo, "ro")
+	}
+
+	source := fmt.Sprintf("%s:%s", clusterIP, defaultExportPath)
+	mounter := mount.New("")
+	err = mounter.Mount(source, targetPath, "nfs", mo)
+	if err != nil {
+		if os.IsPermission(err) {
+			return nil, status.Error(codes.PermissionDenied, err.Error())
+		}
+		if strings.Contains(err.Error(), "invalid argument") {
+			return nil, status.Error(codes.InvalidArgument, err.Error())
+		}
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	return &csi.NodePublishVolumeResponse{}, nil
+}
+
+func (flavor *Flavor) IsMultiWriterVolume(volumeID string) bool {
+	serviceName := fmt.Sprintf("%s%s", serviceNamePrefix, volumeID)
+	_, err := flavor.GetNFSService(serviceName)
+	if err != nil {
+		log.Debugf("unable to obtain service %s volume-id %s", serviceName, volumeID)
+		return false
+	}
+	return true
+}
+
+func (flavor *Flavor) GetMultiWriterClaimFromClaimUID(uid string) (*core_v1.PersistentVolumeClaim, error) {
+	log.Tracef(">>>>> GetMultiWriterClaimFromClaimUid with claim %s", uid)
+	defer log.Tracef("<<<<< GetMultiWriterClaimFromClaimUid")
+
+	claims, err := flavor.claimIndexer.ByIndex("uid", uid)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to retrieve claim by uid %s", uid)
+	} else if len(claims) == 0 {
+		return nil, fmt.Errorf("Requested pvc was not found with uid %s", uid)
+	}
+
+	log.Infof("Found the following claims: %v", claims)
+	return claims[0].(*core_v1.PersistentVolumeClaim), nil
+}
+
+func (flavor *Flavor) getPvFromName(pvName string) (*core_v1.PersistentVolume, error) {
+	log.Tracef(">>>>> getPvFromName with claim %s", pvName)
+	defer log.Tracef("<<<<< getPvFromName")
+
+	pv, err := flavor.kubeClient.CoreV1().PersistentVolumes().Get(pvName, meta_v1.GetOptions{})
+	if err != nil {
+		return nil, err
+	}
+	return pv, nil
+}
+
+func (flavor *Flavor) makeNFSPvc(claim *core_v1.PersistentVolumeClaim) (*core_v1.PersistentVolumeClaim, error) {
+	log.Tracef(">>>>> makeNFSPvc with claim %s", claim.ObjectMeta.Name)
+	defer log.Tracef("<<<<< makeNFSPvc")
+
+	// get copy of original pvc
+	claimClone := claim.DeepCopy()
+	claimClone.ObjectMeta.Namespace = multiWriterNamespace
+	claimClone.ObjectMeta.ResourceVersion = ""
+	// modify name with hpe-multiwriter-pvc-<originalclaim-uid>
+	claimClone.ObjectMeta.Name = fmt.Sprintf("%s-%s", pvcNamePrefix, claim.ObjectMeta.UID)
+	// change access-mode to RWO for underlying nfs pvc
+	claimClone.Spec.AccessModes = []core_v1.PersistentVolumeAccessMode{core_v1.ReadWriteOnce}
+	return claimClone, nil
+}
+
+// Create Kubernetes Persistent Volume Claim
+func (flavor *Flavor) CreateNFSPVC(claim *core_v1.PersistentVolumeClaim) (*core_v1.PersistentVolumeClaim, error) {
+	log.Tracef(">>>>> CreateNFSPVC with claim %s", claim.ObjectMeta.Name)
+	defer log.Tracef("<<<<< CreateNFSPVC")
+
+	newClaim, err := flavor.kubeClient.CoreV1().PersistentVolumeClaims(multiWriterNamespace).Create(claim)
```

Comment:
This uses the same provisioner/storageClass as the incoming request right?

We want to propagate those settings to the volume.

Reply:
yes..all properties are copied with deepcopy, i am overriding only those are needed..i was not sure on destoryOnDelete, do we need to add this or let users decide. I believe if we leave this in offine state, users can still import it later using regular means..